perf(semantic): dereference `parent_id`s to slice only once

overlookmotel opened the pull request:
Small optimization to ancestors iterators. Dereference `&AstNodes` to a slice of parent IDs when creating iterator, rather than dereferencing repeatedly in each call to `next`.

In some cases compiler may have already been able to hoist the dereference out of loops, but where calling code is more complex, it may not.

Boshen commented:
### Merge activity

* **Jul 9, 2:58 AM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

overlookmotel committed:
perf(semantic): dereference `parent_id`s to slice only once (#12137)

Small optimization to ancestors iterators. Dereference `&AstNodes` to a slice of parent IDs when creating iterator, rather than dereferencing repeatedly in each call to `next`.

In some cases compiler may have already been able to hoist the dereference out of loops, but where calling code is more complex, it may not.
